README.md
Generate and configure image avatars
based off of the hash of any arbitrary string content.

This functionality is inspired by GitHub's 
[identicons](https://github.blog/2013-08-14-identicons/),
but uses a different algorithm so will not result in identical outputs.

> **Note**  
> The algorithm of this package may change over time
> to improve generated image quality, 
> so there is no guarantee of stability of generated images between versions. 
> We will consider this a semver break, 
> so if image generation changes without any action by the user, 
> we would for example upgrade the version
> from `0.1.0` -> `0.2.0` or `1.0.0` -> `2.0.0`. 
> We are always open to suggestions about how to improve the algorithm
> and/or image quality so please reach out if you have any suggestions!

## Usage

A simple usage example generating an avatar based off the string `parlough` 
and storing it as a PNG image on the file system.

```dart
import 'dart:io';

import 'package:blockify/blockify.dart';
import 'package:image/image.dart';

/// This example generates a rendered icon using most of the default parameter
/// values but setting the background as transparent. Then it proceeds to
/// encode the result as a PNG with the `image` package, and proceeds to store
/// the encoded PNG in a file named `parlough.png`.
void main() {
  const content = 'parlough';

  // Render the image with a transparent background.
  final image = render(
    content: content,
    backgroundColor: ColorRgba8(0, 0, 0, 0),
  );

  // Encode the image as a PNG image with the `image` package.
  final encoded = encodePng(image);

  // Create a new file with the `dart:io` library.
  final file = File('$content.png');

  // Finally write out the encoded PNG to the newly created file.
  file.writeAsBytesSync(encoded);
}
```

**Generated image:**

![The generated image](https://github.com/parlough/blockify/blob/main/example/parlough.png)

**Another example:**

![Generated with 'dash' string and Dart colors](https://github.com/parlough/blockify/blob/main/example/dash.png)

## Features and bugs

Please file feature requests and bugs in the [issue tracker][tracker].

[tracker]: https://github.com/parlough/blockify/issues


CHANGELOG.md
## 1.0.0

- Update minimum Dart SDK requirement to Dart 3.0
- No longer explicitly name `blockify.dart` library
- Require `package:crypto` 3.0.3 or greater
- **BREAKING**:
  - The primary color generation algorithm has changed, 
    resulting in new outputs if no `primaryColor` is specified
  - Require `package:image` 4.0.17 or greater, 
    which includes v4, a complete overhaul of the package
  - Accept `package:image`'s `Color` objects instead of integers
    for `backgroundColor` and `primaryColor`

## 0.2.0

- Update minimum SDK requirement to Dart 2.17
- Update dependencies to latest versions
- Prepare for use in Dart 3
- Update used linting and analysis options

## 0.1.0

- Initial preview release, supporting null safety.


test/blockify_test.dart
import 'package:blockify/blockify.dart' as blockify;
import 'package:test/test.dart';

const testContent = 'blockify';

void main() {
  group('Test error conditions', () {
    test('Throws if too small blockAmount', () {
      expect(
        () => blockify.render(content: testContent, blockAmount: -5),
        throwsRangeError,
      );
    });

    test('Does not throw with proper blockAmount', () {
      expect(
        () => blockify.render(content: testContent, blockAmount: 8),
        isNot(throwsRangeError),
      );
    });

    test('Throws if too large blockAmount', () {
      expect(
        () => blockify.render(content: testContent, blockAmount: 30),
        throwsRangeError,
      );
    });

    test('Throws if size is negative', () {
      expect(
        () => blockify.render(content: testContent, size: -4),
        throwsRangeError,
      );
    });

    test('Throws if size is 0', () {
      expect(
        () => blockify.render(content: testContent, size: 0),
        throwsRangeError,
      );
    });

    test('Throws if size is not even', () {
      expect(
        () => blockify.render(content: testContent, size: 301),
        throwsArgumentError,
      );
    });

    test('Does not throw if size is even', () {
      expect(
        () => blockify.render(content: testContent, size: 300),
        isNot(throwsArgumentError),
      );
    });
  });

  group('Test image properties', () {
    test('Default size correctly evaluates to 256', () {
      expect(
        blockify.render(content: testContent).height,
        equals(256),
      );
      expect(
        blockify.render(content: testContent).width,
        equals(256),
      );
    });

    test('Can correctly override size', () {
      expect(
        blockify.render(content: testContent, size: 512).height,
        equals(512),
      );
      expect(
        blockify.render(content: testContent, size: 512).width,
        equals(512),
      );
    });
  });
}


example/blockify_example.dart
import 'dart:io';

import 'package:blockify/blockify.dart';
import 'package:image/image.dart';

/// This example generates a rendered icon using most of the default parameter
/// values but setting the background as transparent. Then it proceeds to
/// encode the result as a PNG with the `image` package, and proceeds to store
/// the encoded PNG in a file named `parlough.png`.
///
/// You can find this and some other examples in the repository's
/// `example` directory.
void main() {
  const content = 'parlough';

  // Render the image with a transparent background.
  final image = render(
    content: content,
    backgroundColor: ColorRgba8(0, 0, 0, 0),
  );

  // Encode the image as a PNG image with the `image` package.
  final encoded = encodePng(image);

  // Create a new file with the `dart:io` library.
  final file = File('$content.png');

  // Finally write out the encoded PNG to the newly created file.
  file.writeAsBytesSync(encoded);
}


lib/blockify.dart
/// Generate and render an image based off any [String] you specify
/// with the configurable [render] function
/// to use as profile pictures, identifiers, or more.
library;

import 'dart:convert' show utf8;
import 'dart:math' show max;

import 'package:crypto/crypto.dart' show sha512;
import 'package:image/image.dart';

/// Render an [Image] that is automatically generated
/// based off the SHA-512 of the [content] you specify.
/// As a result, barring any changes in the algorithm between releases,
/// if you specify the same parameters,
/// the resulting image will be the same.
///
/// It will be rendered at a size of [size]*[size]
/// which must be a positive and even integer.
/// This defaults to a `256x256` image.
///
/// It will have the specified [backgroundColor],
/// which defaults to white.
///
/// There will be a square of blocks that is [blockAmount]x[blockAmount]
/// large where each "block" is of equivalent size.
/// This often looks better as an odd number and
/// defaults to `5` and must be no greater than 8.
///
/// These "blocks" will be a single color:
/// either the [primaryColor] you specify
/// or a color generated from the last few bytes of the generated hash.
///
/// You can control the margin outside of the block of "blocks"
/// by specifying a decimal [margin]
/// which if not in the range of 0-1 will be clamped to fit.
Image render({
  required final String content,
  Color? backgroundColor,
  Color? primaryColor,
  final int size = 256,
  final double margin = .12,
  final int blockAmount = 5,
}) {
  if (blockAmount < 1 || blockAmount > 8) {
    throw RangeError.value(
      blockAmount,
      'blockAmount',
      'blockAmount must be between 1 and 8.',
    );
  }

  if (size <= 1) {
    throw RangeError.value(
      size,
      'size',
      'size must be greater than 0.',
    );
  }

  if (size.isOdd) {
    throw ArgumentError.value(
      size,
      'size',
      'The size of the image must be even.',
    );
  }

  // Set the default background color to white.
  backgroundColor ??= ColorUint8.rgba(0xFF, 0xFF, 0xFF, 0xFF);

  final data = sha512.convert(utf8.encode(content)).bytes;

  if (primaryColor == null) {
    final amount = data.length;

    // Take the last 4 bytes as the default primary color.
    primaryColor = ColorUint8.rgba(
      data[amount - 3 - 1],
      data[amount - 2 - 1],
      data[amount - 1 - 1],
      data[amount - 0 - 1],
    );
  }

  final maxChannels = max(primaryColor.length, backgroundColor.length);

  final image = Image(
    width: size,
    height: size,
    numChannels: maxChannels,
    backgroundColor: backgroundColor,
  );

  image.clear(backgroundColor);

  final marginSize = (size * margin.clamp(0, 1)).truncate();
  final blockSize = (size - (marginSize * 2)) ~/ blockAmount;

  void drawCellRect(int x1, int y1, final Color color) {
    x1 = x1 * blockSize + marginSize;
    y1 = y1 * blockSize + marginSize;
    final x2 = x1 + blockSize - 1;
    final y2 = y1 + blockSize - 1;
    for (var x = x1; x <= x2; x++) {
      for (var y = y1; y <= y2; y++) {
        image.setPixel(x, y, color);
      }
    }
  }

  final evenMiddle = blockAmount.isEven;
  final columns = (blockAmount / 2).ceil();

  for (var x = 0; x < columns; x++) {
    for (var y = 0; y < blockAmount; y++) {
      // Choose to include the block if its byte is odd.
      if (data[x * blockAmount + y].isEven) continue;

      drawCellRect(x, y, primaryColor);
      if (x == columns - 1) {
        if (evenMiddle) {
          drawCellRect(x + 1, y, primaryColor);
        }
      } else {
        drawCellRect(blockAmount - x - 1, y, primaryColor);
      }
    }
  }

  return image;
}
